ui/packages/catalog-ui-search/src/main/webapp/react-component/menu/menu.js

```diff
@@ -67,6 +67,12 @@ class Menu extends React.Component {
   render() {
     const { value, children } = this.props
 
+    if (typeof this.state.active === 'undefined' && children.length > 0) {
```

Comment:
We shouldn't be mutating state in this way, we should stick to setState.  However, setState in render is something we shouldn't be doing.  

Instead we should move this into componentDidUpdate and do something like this:
```
componentDidUpdate(prevProps) {
    if (prevProps.children !== this.props.children) {
      if (
        typeof this.state.active === 'undefined' &&
        this.props.children.length > 0
      ) {
        this.setState({
          active: this.props.children[0].props.value,
        })
      } else if (this.props.children.length === 0) {
        this.setState({
          active: undefined,
        })
      }
    }
  }
```
Doesn't have to be exactly that, but something similar.

Reply:
That makes sense except for one issue that I see in [the documentation of `componentDidUpdate`](https://reactjs.org/docs/react-component.html#componentdidupdate):

> This method is not called for the initial render.

How can I make this update the state on the initial render as well when the children of the menu are populated? Otherwise, I believe the first item in the menu will not be selected until the menu is rendered the second time.

Reply:
We actually need to test what effect on performance this will have, I suspect due to the nature of this component and how it is used it would not be a concern.   Did your tests reveal any negative side-effects?

Reply:
I have not seen any changes in performance. I did note when the `componentDidUpdate` and `render` methods were called and everything seems to be run according to the React docs.

I do notice that the `componentDidUpdate` is called twice per render when the state is changed in the call, the second call always finding the initial condition with `prevProps` to be false. This is expected behavior. Two lists are being compared there, so I could see potential for some performance impact under very specific circumstances that don't currently exist in the UI like two large lists of complex objects that are almost identical. I don't know how much of an impact that would make and the case seems rare (and currently nonexistent).

Reply:
In response to [this comment](https://github.com/codice/ddf/pull/4075#discussion_r237150976), it looks like initializing state in the constructor will fill that gap for the first render. I reworked the `active` initialization logic in the constructor to reflect the change.